File: macros/PrintYields.cc
// STL
#include <iostream>
#include <fstream>

// ROOT
#include "TString.h"
#include "TFile.h"
#include "TTree.h"

// Cuts
#include "Cuts2017.hh"
//#include "Cuts2016.hh"

using namespace std;


/*
**  PrintYields
**
**  Prints event yields for each channel.
*/

void PrintYields()
{

    //
    //  SAMPLE INFO
    //

    const unsigned N = 8;   // Channel indices
    unsigned                LL = 0, MM = 1, EE = 2, L4 = 3, M4 = 4, ME = 5, EM = 6, E4 = 7;
    TString selection[N] = {"ll",   "mumu", "ee",   "4l",   "4m",   "2m2e", "2e2m", "4e"};
    unsigned chanIdx[N] = { 2,      3,      4,      5,      6,      7,      8,      9};






    ////
    ////
    ////    READ TREES
    ////
    ////


    TString inPath  = EOS_PATH + "/Selected/" + YEAR_STR + "/";
    TString prefix = "selected";



    //
    //  DATA
    //

    float data[N];

    // Muon file
    TString muName = inPath + prefix + "_" + MU_SUFF + ".root";
    TFile *muFile = TFile::Open(muName);
    cout << "Opened " << muName << endl;

    // Electron file
    TString elName = inPath + prefix + "_" + EL_SUFF + ".root";
    TFile *elFile = TFile::Open(elName);
    cout << "Opened " << elName << endl;

    for (unsigned i = 0; i < N; i++)
    {
        TTree *tree;

        if      ((i == LL) || (i == L4))
            continue;
        else if ((i == MM) || (i == M4) || (i == ME))
            muFile->GetObject(selection[i] + "_" + MU_SUFF, tree);
        else if ((i == EE) || (i == E4) || (i == EM))
            elFile->GetObject(selection[i] + "_", tree);

        data[i] = tree->GetEntries();
        cout << data[i] << ", ";
    }
    cout << endl;

    muFile->Close();
    elFile->Close();



/*
    //
    //  MONTE CARLO
    //

    TH1 *mcHist[N_MC];
    for (unsigned j = 0; j < N_MC; j++) // sample loop
    {
        TString inName = inPath + prefix + "_" + MC_SUFF[j] + ".root";
        TFile *inFile = TFile::Open(inName);

        cout << "Opened " << inName << endl;

        TH1 *hist_;
        inFile->GetObject("SelectedEvents_" + MC_SUFF[j], hist_);
        hist_->SetDirectory(0);
        hist_->Sumw2();

        mcHist[j] = hist_;

        inFile->Close();
    }






    ////
    ////
    ////    CALCULATIONS
    ////
    ////


    //
    //  SCALING
    //

    // Make a histogram of the appropriate lumi for each channel
    TH1D *lumiHist = new TH1D("Lumi", "Lumi", 10, 0.5, 10.5);

    lumiHist->SetBinContent(chanIdx[MM], MUON_TRIG_LUMI);
    lumiHist->SetBinContent(chanIdx[M4], MUON_TRIG_LUMI);
    lumiHist->SetBinContent(chanIdx[ME], MUON_TRIG_LUMI);

    lumiHist->SetBinContent(chanIdx[EE], ELEC_TRIG_LUMI * ELEC_TRIG_SF);
    lumiHist->SetBinContent(chanIdx[EM], ELEC_TRIG_LUMI * ELEC_TRIG_SF);
    lumiHist->SetBinContent(chanIdx[E4], ELEC_TRIG_LUMI * ELEC_TRIG_SF);

    
    // Scale by appropriate lumi, etc.
    for (unsigned j = 0; j < N_MC; j++) // sample loop
    {
        float sf = 1000 * XSEC[j] / NGEN[j];
        mcHist[j]->Multiply(lumiHist);
        mcHist[j]->Scale(sf);
    }



    //
    //  SUMS
    //

    TH1 *mcTotal;
    mcTotal = (TH1*) mcHist[0]->Clone("ExpectedEvents");
    for (unsigned j = 1; j < N_MC; j++)
        mcTotal->Add(mcHist[j]);

    TH1 *sigBackground = (TH1*) mcHist[DY]->Clone("BackgroundEvents");

    TH1 *dyBackground = (TH1*) mcHist[ZZ]->Clone("DYBackgroundEvents");
    for (unsigned j = 2; j < N_MC; j++)
    {
        sigBackground->Add(mcHist[j]);
        dyBackground->Add(mcHist[j]);
    }
    for (unsigned i = MM; i < L4; i++)
        sigBackground->SetBinContent(chanIdx[i], 0);
    for (unsigned i = M4; i < N; i++)
        dyBackground->SetBinContent(chanIdx[i], 0);

    TH1 *obsMinusBg = (TH1*) dataHist->Clone("ObsMinusBgEvents");
    obsMinusBg->Add(sigBackground, -1);
    obsMinusBg->Add(dyBackground, -1);




    //
    //  LATEX
    //

    TString selLaTeX[N] = {"\\lplm", "\\mumu", "\\ee", "4\\ell", "4\\mu", "2\\mu2\\el","","4\\el"};
    for (unsigned i = MM; i < N; i++)
    {
        if (i == L4)
            continue;
        if (i == EM)
            continue;

        unsigned SIG = (i < L4) ? DY : ZZ;


        int nObserved = dataHist->GetBinContent(chanIdx[i]);
        float nExpected = mcTotal->GetBinContent(chanIdx[i]);
        float uExpected = mcTotal->GetBinError(chanIdx[i]);
        float nSignal = mcHist[SIG]->GetBinContent(chanIdx[i]);
        float uSignal = mcHist[SIG]->GetBinError(chanIdx[i]);
        float nBackground = (i > L4) ? sigBackground->GetBinContent(chanIdx[i])
                                     : dyBackground->GetBinContent(chanIdx[i]);
        float uBackground = (i > L4) ? sigBackground->GetBinError(chanIdx[i])
                                     : dyBackground->GetBinError(chanIdx[i]);

        if (i == ME)
        {
            nObserved += dataHist->GetBinContent(chanIdx[EM]);
            nExpected += mcTotal->GetBinContent(chanIdx[EM]);
            uExpected = sqrt(pow(uExpected, 2) + pow(mcTotal->GetBinError(chanIdx[EM]), 2));
            nSignal += mcHist[ZZ]->GetBinContent(chanIdx[EM]);
            uSignal = sqrt(pow(uSignal, 2) + pow(mcHist[ZZ]->GetBinError(chanIdx[EM]), 2));
            nBackground += sigBackground->GetBinContent(chanIdx[EM]);
            uBackground = sqrt(pow(uBackground,2)+pow(sigBackground->GetBinError(chanIdx[EM]),2));
        }
        if (i < L4)
        {
            nObserved = round(nObserved);
            nExpected = round(nExpected);       uExpected = round(uExpected);
            nSignal = round(nSignal);           uSignal = round(uSignal);
            nBackground = round(nBackground);   uBackground = round(uBackground);
        }

        TString texName = YEAR_STR + "_" + selection[i] + ".tex";

        ofstream texFile;
        texFile.open(texName);
        if (i > L4)
        {
            texFile.precision(2);
            texFile << fixed;
            texFile << "\\begin{tabular}{l l l S[table-format=3.2] l S[table-format=2.2]}" << endl; 
        }
        else
        {
            texFile.precision(8);
            texFile << "\\sisetup{group-minimum-digits=4}" << endl << endl;
            texFile << "\\begin{tabular}{l l l S[table-format=8.0] l S[table-format=5.0]}" << endl; 
        }

        texFile << "\\toprule" << endl << "\t\\multicolumn{3}{l}{Type} & ";
        texFile << "\\multicolumn{3}{l}{$N_{" << selLaTeX[i] << "}$ (events)} \\\\" << endl;
        texFile << "\\midrule" << endl;
        texFile << "\t\\multicolumn{3}{l}{Observed} & " << nObserved << " \\\\" << endl;
        texFile << "\\addlinespace\\addlinespace" << endl;
        texFile << "\t\\multicolumn{3}{l}{Expected} & ";
        texFile << nExpected << " & $\\pm$ & " << uExpected << " \\\\" << endl;
        texFile << "\\addlinespace" << endl << "\t& \\multicolumn{2}{l}{Signal} & ";
        texFile << nSignal << " & $\\pm$ & " << uSignal << " \\\\" << endl;
        texFile << "\\addlinespace" << endl << "\t& \\multicolumn{2}{l}{Background} & ";
        texFile << nBackground << " & $\\pm$ & " << uBackground << " \\\\" << endl;
        texFile << "\\addlinespace" << endl;

        for (unsigned j = ZZ; j < N_MC; j++)
        {
            if (j == SIG)
                continue;

            float nSample = mcHist[j]->GetBinContent(chanIdx[i]);
            float uSample = mcHist[j]->GetBinError(chanIdx[i]);

            if (i < L4)
            {
                nSample = round(nSample);       uSample = round(uSample);
            }

            texFile << "\t\t& & $" << MC_TEX[j] << "$ & ";
            texFile << nSample << " & $\\pm$ & " << uSample << " \\\\" << endl;
        }
        texFile << "\\bottomrule" << endl << "\\end{tabular}" << endl;

        texFile.close();
        cout << "Wrote LaTeX table to " << texName << endl;
    }
*/
}

File: python/PrintYields.py
from __future__ import print_function
from __future__ import division

import numpy as np

from ROOT import TFile, TTree, TH1D

#from Cuts2018 import *
from Cuts2017 import *
#from Cuts2016 import *
#from Cuts2012 import *



##
##  SAMPLE INFO
##

selection   = ["mumu", "ee", "4l", "4m", "2m2e", "4e"]
selTeX      = { "mumu":r"\MM",  "ee":r"\EE",
                "4l":r"4\Pell", "4m":r"4\PGm",  "4e":r"4\Pe",   "2m2e":r"2\PGm 2\Pe"
                }
selDef      = { "mumu":"MM",    "ee":"EE",  "4l":"4L",  "4m":"4M",  "4e":"4E",  "2m2e":"2M2E"   }
T = np.dtype([(sel, 'f4') for sel in selection])



##
##  DATA
##

inPath = EOS_PATH + "/Selected/" + YEAR_STR + "_new/"
prefix = "selected"

# Muon file
muName = prefix + "_" + MU_SUFF + ".root"
muFile = TFile.Open(inPath + muName)
print("Opened", inPath + muName)

# Electron file
elName = prefix + "_" + EL_SUFF + ".root"
elFile = TFile.Open(inPath + elName)
print("Opened", inPath + elName)

# Get yields
data = np.zeros(1, dtype=T)

for sel in selection:
    muTree = muFile.Get(sel + "_" + MU_SUFF)
    elTree = elFile.Get(sel + "_" + EL_SUFF)
    data[sel] = muTree.GetEntries() + elTree.GetEntries()

muFile.Close()
elFile.Close()



##
##  MONTE CARLO
##

mc_arr, mc_unc_arr = np.zeros(N_MC, dtype=T), np.zeros(N_MC, dtype=T)
sig, sig_unc = np.zeros(1, dtype=T), np.zeros(1, dtype=T)
mc, mc_unc = {}, {}
row = 0

# Loop over all samples
for suff in MC_SUFF:
    inName = prefix + "_" + suff + ".root"
    inFile = TFile.Open(inPath + inName)
    print("Opened", inPath + inName)

    # Get histograms
    for sel in selection:
        if suff in ["zjets_m-50", "ttbar", "tt_2l2nu"] and sel in ["4l", "4m", "2m2e", "4e"]:
            continue
        
        sf = INT_LUMI * 1000 * XSEC[suff] / NGEN[suff]

        cut = ""
        weight = "weight"

        hist = TH1D("hist", "", 1, 0, 2)
        tree = inFile.Get(sel + "_" + suff)

        
        # Get signal
        if (suff == "zz_4l" and sel in ["4l", "4m", "2m2e", "4e"]) or (suff == "zjets_m-50" and sel in ["mumu", "ee"]):
            tree.Draw("1>>hist", "!hasTauDecay * " + weight, "goff")
            sig[sel] = sf * hist.Integral()
            sig_unc[sel] = sf * np.sqrt(tree.GetEntries("!hasTauDecay"))
            cut = "hasTauDecay"
            weight = cut + " * " + weight
 
        tree.Draw("1>>hist", weight, "goff")

        mc_arr[row][sel] = sf * hist.Integral()
        mc_unc_arr[row][sel] = sf * np.sqrt(tree.GetEntries(cut))

        mc_arr[row][sel] = np.abs(mc_arr[row][sel])

        hist.Delete()

    mc[suff] = mc_arr[row]
    mc_unc[suff] = mc_unc_arr[row]
    row = row + 1
    inFile.Close()



##
##  ADD SAMPLES
##

# Take average of ttbar (inclusive) and tt_2l2nu
if YEAR_STR != "2012":
    for sel in ["mumu", "ee"]:
        mc['ttbar'][sel] = mc['ttbar'][sel] + mc['tt_2l2nu'][sel]
        mc_unc['ttbar'][sel] = np.sqrt(mc_unc['ttbar'][sel] ** 2 + mc_unc['tt_2l2nu'][sel] ** 2)
        mc['tt_2l2nu'][sel] = 0
        mc_unc['tt_2l2nu'][sel] = 0


# Get nonprompt background
infile = "nonprompt" + YEAR_STR + ".npz"
npzfile = np.load(infile)
npt_, npt_unc_ = npzfile['npt'], npzfile['npt_unc']
npt, npt_unc = np.zeros(1, dtype=T), np.zeros(1, dtype=T)

for sel in ["4l", "4m", "2m2e", "4e"]:
    npt[sel] = npt_[sel]
    npt_unc[sel] = npt_unc_[sel]


# Get total expected and background events
exp, exp_unc = np.zeros(1, dtype=T), np.zeros(1, dtype=T)
bg, bg_unc = np.zeros(1, dtype=T), np.zeros(1, dtype=T)
pur, pur_unc = np.zeros(1, dtype=T), np.zeros(1, dtype=T)
sf, sf_unc = np.zeros(1, dtype=T), np.zeros(1, dtype=T)

for sel in selection:
    exp[sel]        = np.sum(mc_arr[sel]) + sig[sel] + npt[sel]
    exp_unc[sel]    = np.sqrt(np.sum(mc_unc_arr[sel] ** 2) + sig_unc[sel] ** 2 + npt_unc[sel] ** 2)
    bg[sel]         = np.sum(mc_arr[sel]) + npt[sel]
    bg_unc[sel]     = np.sqrt(np.sum(mc_unc_arr[sel] ** 2) + npt_unc[sel] ** 2)
    pur[sel]        = sig[sel] / exp[sel] * 100
    pur_unc[sel]    = pur[sel] * np.sqrt(sig_unc[sel] / sig[sel] ** 2 + exp_unc[sel] / exp[sel] ** 2)
    sf[sel]         = data[sel] / exp[sel] * 100
    sf_unc[sel]     = sf[sel] * np.sqrt(1 / data[sel] + exp_unc[sel] / exp[sel] ** 2)



##
##  WRITE TEX FILES
##

fmt = '%.2f'

fileName = "Yield" + YEAR_STR + ".tex"
f = open(fileName, "w")

f.write(r"\begin{tabular}{lll r@{ $\pm$ }r r@{ $\pm$ }r r@{ $\pm$ }r r@{ $\pm$ }r r@{ $\pm$ }r r@{ $\pm$ }r}" + "\n")
f.write(r"\toprule" + "\n")

f.write("\t" + r"\multicolumn{3}{l}{Type}")
for sel in selection:
    f.write(r" & \multicolumn{2}{l}{$N_{" + selTeX[sel] + r"}$}")
f.write(r" \\" + "\n")

f.write(r"\midrule" + "\n")

f.write("\t" + r"\multicolumn{3}{l}{Observed}")
for sel in selection:
    f.write(r" & \multicolumn{2}{l}{" + '%.0f' % np.squeeze(data[sel]) + "}")
f.write(r" \\" + "\n")

f.write(r"\addlinespace\addlinespace" + "\n")

f.write("\t" + r"\multicolumn{3}{l}{Expected}")
for sel in selection:
    f.write(" & " + fmt % np.squeeze(exp[sel]) + " & " + fmt % np.squeeze(exp_unc[sel]))
f.write(r" \\" + "\n")

f.write(r"\addlinespace" + "\n")

f.write("\t&\t" + r"\multicolumn{2}{l}{Signal}")
for sel in selection:
    f.write(" & " + fmt % np.squeeze(sig[sel]) + " & " + fmt % np.squeeze(sig_unc[sel]))
f.write(r" \\" + "\n")

f.write(r"\addlinespace" + "\n")

f.write("\t&\t" + r"\multicolumn{2}{l}{Background}")
for sel in selection:
    f.write(" & " + fmt % np.squeeze(bg[sel]) + " & " + fmt % np.squeeze(bg_unc[sel]))
f.write(r" \\" + "\n")

f.write(r"\addlinespace" + "\n")

f.write("\t&\t&\t" + r"Nonprompt & \multicolumn{4}{c}{}")
for sel in ["4l", "4m", "2m2e", "4e"]:
    f.write(" & " + fmt % np.squeeze(npt[sel]) + " & " + fmt % np.squeeze(npt_unc[sel]))
f.write(r" \\" + "\n")

for suff in MC_SUFF:
    if suff == "tt_2l2nu":
        continue
    else:
        f.write("\t&\t&\t" + MC_TEX[suff])

        for sel in selection:
            if suff in ["zjets_m-50", "ttbar"] and sel in ["4l", "4m", "2m2e", "4e"]:
                continue
            else:
                f.write(" & " + fmt % np.squeeze(mc[suff][sel]) + " & "
                        + fmt % np.squeeze(mc_unc[suff][sel]))
        f.write(r" \\" + "\n")

f.write(r"\midrule" + "\n")

f.write("\t" + r"\multicolumn{3}{l}{Purity (\%)}")
for sel in selection:
    f.write(" & " + fmt % np.squeeze(pur[sel]) + " & " + fmt % np.squeeze(pur_unc[sel]))
f.write(r" \\" + "\n")

f.write(r"\addlinespace" + "\n")

f.write("\t" + r"\multicolumn{3}{l}{Exp. $\mapsto$ Obs. (\%)}")
for sel in selection:
    f.write(" & " + fmt % np.squeeze(sf[sel]) + " & " + fmt % np.squeeze(sf_unc[sel]))
f.write(r" \\" + "\n")

f.write(r"\bottomrule" + "\n")
f.write(r"\end{tabular}" + "\n")

f.close()
print("Wrote table to", fileName)



##
##  SAVE
##

print("")
outfile = "yields" + YEAR_STR + ".npz"
np.savez(outfile, data=data, exp=exp, exp_unc=exp_unc, sig=sig, sig_unc=sig_unc, bg=bg,
        bg_unc=bg_unc, mc=mc_arr, mc_unc=mc_unc_arr, npt=npt, npt_unc=npt_unc)

print("Wrote arrays to", outfile)
